### PR TITLE
Migrate to JDK 11 and Java 11 syntax

### PR DESCRIPTION
**Pull request by armiol**
This changeset migrates the project to JDK 11.

In particular, the changes were made to both build scripts and the source code.

**Java 11 Target**

Compilation target is now set to Java 11 for both the project source code and `buildSrc`.

Please note that the Kotlin sources in the `buildSrc` are still compiled to Java 8. This is because of the compiler shipped along with the `kotlin-dsl` plugin, which isn't capable of configuring the compilation target. See the [corresponding issue](https://youtrack.jetbrains.com/issue/KT-47520) in Kotlin, which has been already addressed in 1.6.0 — but not in Gradle's `kotlin-dsl`.

**`var`**

Implicit `var` syntax is now used instead of explicit type declaration where appropriate. 

In fact, we tend to write our code in a way that makes `var`s suitable in almost 100% of cases. 
In scope of this changeset, it's up to reviewers to judge whether such an approach is still comprehensible and readable. But the final policy on `var` usage is yet to be defined.

**`Javac` Extension**

The `Javac` extensions which we have in `buildSrc` are modified to exclude that JDK8-only restriction that we had previously. This is only done in scope of this repository. Eventually, this restriction will have to be removed in `config`.

**Error Prone**

Error Prone's `javac` module is no longer enabled, as no custom compiler is required for Error Prone starting JDK 9.

Minor changes:

* The GitHub Actions workflows are modified to run on JDK 11. This change has also been done locally. And similar to other `config`-related updates, that will be eventually moved to `config` repository.

* The `README.md` has been updated with the JDK information — and! — the new badge, which displays the current GitHub Actions status.

* The latest `config` contents were used, so that the build scripts had to be updated to `config`'s latest API.

The version of the library is now set to `2.0.0-SNAPSHOT.63`.

### CODE REVIEW

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/github/pages/UpdateGitHubPages.kt [hunk 1]
```diff
@@ -92,254 +88,115 @@
...
+    private fun TaskContainer.registerCopyJavadoc(allowInternalJavadoc: Boolean) {
+        val inputs = composeInputs(allowInternalJavadoc)
+        register(copyJavadoc, Copy::class.java) {
             doLast {
-                try {
-                    updateGhPages(checkoutTempFolder, project, javadocOutputPath)
-                } finally {
-                    cleanup()
-                }
+                from(*inputs.toTypedArray())
```
[the remark this answers is not in the document]
Comment (armiol) [reply buildSrc/src/main/kotlin/io/spine/internal/gradle/github/pages/UpdateGitHubPages.kt:151]: @sonatype-lift ignore

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/github/pages/Update.kt [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        if (!sshConfigFile.exists()) {
+            val parentDir = sshConfigFile.canonicalFile.parentFile
+            parentDir.mkdirs()
+            sshConfigFile.createNewFile()
+        }
+        return sshConfigFile
+    }
+
+    /** Executes a command in the project [rootFolder]. */
+    private fun execute(vararg command: String): String = Cli(rootFolder).execute(*command)
```
[the remark this answers is not in the document]
Comment (armiol) [reply buildSrc/src/main/kotlin/io/spine/internal/gradle/github/pages/Update.kt:196]: @sonatype-lift ignore

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 3]
```diff
@@ -164,22 +161,22 @@ public BoundedContextBuilder configureTenantIndex(BoundedContextBuilder builder)
     createRecordStorage(ContextSpec context, RecordSpec<I, R, ?> spec) {
         checkNotNull(context);
         checkNotNull(spec);
-        StorageConfiguration<I, R> config = configurationWith(context, spec);
-        Optional<CreateStorage<I, R>> custom = customStorages.find(spec);
-        RecordStorage<I, R> result =
+        var config = configurationWith(context, spec);
+        var custom = customStorages.find(spec);
+        var result =
                 custom.map(callback -> callback.apply(config))
```
Comment (dmdashenkov) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java:167]: Please join this line to the previous one.

File: datastore/src/main/java/io/spine/server/storage/datastore/DsQueryIterator.java [hunk 4]
```diff
@@ -106,8 +106,8 @@ private void terminate() {
      * query results.
      */
     StructuredQuery<R> nextPageQuery() {
-        Cursor cursorAfter = currentPage.getCursorAfter();
-        StructuredQuery<R> queryForMoreResults =
+        var cursorAfter = currentPage.getCursorAfter();
+        var queryForMoreResults =
                 query.toBuilder()
```
Comment (dmdashenkov) [inline datastore/src/main/java/io/spine/server/storage/datastore/DsQueryIterator.java]: Same as above.

File: datastore/src/main/java/io/spine/server/storage/datastore/delivery/SetNodeIfAbsent.java [hunk 5]
```diff
@@ -60,13 +60,12 @@
...
-                        .setIndex(index)
-                : previous.toBuilder();
+        var builder = previous == null
+                      ? ShardSessionRecord.newBuilder()
+                              .setIndex(index)
+                      : previous.toBuilder();
 
-        ShardSessionRecord updated =
+        var updated =
                 builder.setPickedBy(nodeToSet)
```
Comment (dmdashenkov) [inline datastore/src/main/java/io/spine/server/storage/datastore/delivery/SetNodeIfAbsent.java]: Same here.

File: datastore/src/main/java/io/spine/server/storage/datastore/query/ToRecords.java [hunk 6]
```diff
@@ -66,12 +66,13 @@
...
-        Stream<Entity> stream =
+        var entities = result.entities();
+        var stream =
                 entities.stream()
                         .filter(Objects::nonNull);
         stream = filter(stream);
-        ImmutableList<R> records =
+        @SuppressWarnings("ConstantConditions") /* `null` were already filtered out. */
+        var records =
                 stream.map(this::toRecord)
```
Comment (dmdashenkov) [inline datastore/src/main/java/io/spine/server/storage/datastore/query/ToRecords.java:74]: Same in this file.

File: datastore/src/main/java/io/spine/server/storage/datastore/tenant/NamespaceConverters.java [hunk 7]
```diff
@@ -135,19 +135,21 @@ private DomainNamespaceConverter() {
...
         @Override
         protected TenantId significantStringToTenantId(String namespace) {
-            InternetDomain domain = InternetDomain.newBuilder()
-                    .setValue(namespace)
-                    .vBuild();
-            TenantId tenantId = TenantId.newBuilder()
-                    .setDomain(domain)
-                    .vBuild();
+            var domain =
+                    InternetDomain.newBuilder()
```
Comment (dmdashenkov) [inline datastore/src/main/java/io/spine/server/storage/datastore/tenant/NamespaceConverters.java]: Same here. I don't see why did we split this statement in such a way.

Comment (dmdashenkov) [reply datastore/src/main/java/io/spine/server/storage/datastore/tenant/NamespaceConverters.java]: There are several similar cases in this file.

File: datastore/src/test/java/io/spine/server/storage/datastore/DatastoreWrapperTest.java [hunk 8]
```diff
@@ -308,34 +306,37 @@ void testIaeOnBatchReadWithZeroSize() {
...
-                               .setValue(tenantId1)
-                               .vBuild();
-        TenantId id2 = TenantId.newBuilder()
-                               .setEmail(emailAddress2)
-                               .vBuild();
-        TenantId id3 = TenantId.newBuilder()
-                               .setDomain(internetDomain3)
-                               .vBuild();
+        var id1 = TenantId
+                .newBuilder()
```
Comment (dmdashenkov) [inline datastore/src/test/java/io/spine/server/storage/datastore/DatastoreWrapperTest.java]: I think, our current convention dictates that the `newBuilder()` was on the same line with the type name.

File: datastore/src/test/java/io/spine/server/storage/datastore/tenant/NamespaceIndexTest.java [hunk 9]
```diff
@@ -218,49 +215,49 @@ private static void testSynchronizeAccessMethods() throws InterruptedException {
...
         };
 
         // Test execution threads
-        Thread firstThread = new Thread(flow);
-        Thread secondThread = new Thread(flow);
+        var firstThread = new Thread(flow);
+        var secondThread = new Thread(flow);
 
         // Collect thread failures
         Map<Thread, Throwable> threadFailures = new HashMap<>(2);
```
Comment (dmdashenkov) [inline datastore/src/test/java/io/spine/server/storage/datastore/tenant/NamespaceIndexTest.java]: We should consider
```java
var threadFailures = new HashMap<Thread, Throwable>(2);
```
Seems a bit more concise.
However, the next line with the method reference cast to a functional interface looks just right without the `var` to me.

**Review by dmdashenkov — APPROVED** [review]
LGTM with a suggestion for further `var`-ification and a bunch of suggestions on code alignment.

File: .idea/misc.xml [hunk 10]
```diff
@@ -38,5 +38,5 @@
       </value>
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK" />
-</project>
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" project-jdk-name="1.8" project-jdk-type="JavaSDK" />
```
Comment (alexander-yevsyukov) [inline .idea/misc.xml]: Why does the value of `project-jdk-name` stay `1.8`?

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 11]
```diff
@@ -269,7 +266,7 @@ final DatastoreWrapper wrapperFor(ContextSpec spec) {
     @Internal
     public final DatastoreWrapper
     systemWrapperFor(Class<? extends Storage<?, ?>> targetStorage, boolean multitenant) {
-        DatastoreWrapper wrapper = sysWrappers
+        var wrapper = sysWrappers
                 .computeIfAbsent(targetStorage, k -> newDatastoreWrapper(multitenant));
```
Comment (alexander-yevsyukov) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java]: Please join this line with the previous one.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 12]
```diff
@@ -303,8 +300,8 @@ public static Builder newBuilder() {
     @VisibleForTesting
     public static Builder newBuilderWithDefaults(Datastore datastore) {
         checkNotNull(datastore);
-        Builder result = newBuilder().setDatastore(datastore)
-                                     .withDefaults();
+        var result = newBuilder().setDatastore(datastore)
```
Comment (alexander-yevsyukov) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java]: Consider moving `.setDatastore()` to the following line to reduce indentation of the block.

File: datastore/src/main/java/io/spine/server/storage/datastore/query/DsFilters.java [hunk 13]
```diff
@@ -157,15 +156,15 @@ private static void handleParameters(ImmutableList<SubjectParameter<?, ?, ?>> pa
         checkState(predicate.children()
                             .isEmpty(),
                    "Children collection must be empty for a conjunctive predicate group.");
-        ImmutableList<SubjectParameter<?, ?, ?>> parameters = predicate.allParams();
+        var parameters = predicate.allParams();
         List<StructuredQuery.Filter> filters =
                 parameters.stream()
```
Comment (alexander-yevsyukov) [inline datastore/src/main/java/io/spine/server/storage/datastore/query/DsFilters.java]: Please have `parameters.stream()` on the previous line here and above.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java (hunk 3 above)
Comment (armiol) [reply datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java:167]: Left as-is to preserve `.map` to `.orElse` alignment in IDEA.

File: datastore/src/main/java/io/spine/server/storage/datastore/query/ToRecords.java (hunk 6 above)
Comment (armiol) [reply datastore/src/main/java/io/spine/server/storage/datastore/query/ToRecords.java:74]: Moved to a single line: 
```java
 entities.stream().filter(Objects::nonNull);
```

As there is another `.filter(..)` on the next line.

File: datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java [hunk 14]
```diff
@@ -118,16 +120,16 @@ protected void writeAllRecords(Iterable<? extends RecordWithColumns<I, R>> recor
...
-        ImmutableList<Entity> prepared = entitiesToWrite.build();
+        var prepared = entitiesToWrite.build();
         write((storage) -> datastore.createOrUpdate(prepared));
     }
 
     @Override
     protected Iterator<R> readAllRecords(RecordQuery<I, R> query) {
-        Iterable<R> result =
+        var result =
                 read((storage) -> DsLookup.onTopOf(datastore, columnFilterAdapter, dsSpec)
```
Comment (alexander-yevsyukov) [inline datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java]: How about this layout?

```java
var result = read((storage) ->
        DsLookup.onTopOf(datastore, columnFilterAdapter, dsSpec)
            .with(query)
            .execute()
        );
```

File: datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java [hunk 15]
```diff
@@ -160,27 +162,27 @@ public final Kind kind() {
...
+        var id = recordWithCols.id();
+        var key = keyOf(id);
+        var entity = builderFromMessage(record, key);
 
         recordWithCols.columnNames()
                       .forEach(columnName -> {
-                          Value<?> columnValue = recordWithCols.columnValue(columnName,
-                                                                            columnMapping);
+                          var columnValue = recordWithCols.columnValue(columnName,
+                                                                       columnMapping);
```
Comment (alexander-yevsyukov) [inline datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java]: Can we put `columnMapping);` on the previous line now?

File: datastore/src/test/java/io/spine/server/storage/datastore/query/DsFiltersTest.java [hunk 16]
```diff
@@ -99,26 +92,26 @@ void testCompositeEither() {
...
                                   project -> project.dueDate().isLessThan(dueDateValue))
                           .build();
 
-        FilterAdapter adapter = FilterAdapter.of(new DsColumnMapping());
-        Subject<StgProjectId, StgProject> subject = query.subject();
-        Collection<StructuredQuery.Filter> filters =
+        var adapter = FilterAdapter.of(new DsColumnMapping());
+        var subject = query.subject();
+        var filters =
                 fromPredicate(subject.predicate(), adapter);
```
Comment (alexander-yevsyukov) [inline datastore/src/test/java/io/spine/server/storage/datastore/query/DsFiltersTest.java]: Please have it on the previous line.

File: datastore/src/test/java/io/spine/server/storage/datastore/tenant/NamespaceWithCustomConverterTest.java [hunk 17]
```diff
@@ -63,23 +62,23 @@ void testFromTenantId() {
     @Test
     @DisplayName("construct from `Key`")
     void testFromKey() {
-        String ns = "my.test.namespace.from.key";
-        Key key = Key.newBuilder(PROJECT_ID.value(), "some.kind", ns)
-                     .build();
-        Namespace namespace = Namespace.fromNameOf(key, true, factory);
+        var ns = "my.test.namespace.from.key";
+        var key = Key
+                .newBuilder(PROJECT_ID.value(), "some.kind", ns)
```
Comment (alexander-yevsyukov) [inline datastore/src/test/java/io/spine/server/storage/datastore/tenant/NamespaceWithCustomConverterTest.java]: Please have this on the previous line.

**Review by alexander-yevsyukov — APPROVED** [review]
LGTM, also have a look at my comments.

File: datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java (hunk 14 above)
Comment (armiol) [reply datastore/src/main/java/io/spine/server/storage/datastore/record/DsRecordStorage.java]: Here is what I did:

```java
    @Override
    protected Iterator<R> readAllRecords(RecordQuery<I, R> query) {
        var result = read((storage) -> lookupWith(query).execute());
        return result.iterator();
    }

    @NonNull
    private PreparedQuery<I, R> lookupWith(RecordQuery<I, R> query) {
        return DsLookup.onTopOf(datastore, columnFilterAdapter, dsSpec)
                       .with(query);
    }
```

**Pull request merged**